app/controllers/fhir/r4/api_controller.rb

```diff
@@ -342,12 +340,115 @@ def create
                             created_by: @current_actor_label,
                             comment: "New vaccine added via API (ID: #{resource.id}).")
       end
+    when 'observation'
+      return if doorkeeper_authorize!(*OBSERVATION_WRITE_SCOPES)
+
+      resource, fhir_map = build_laboratory(contents)
+
+      ActiveRecord::Base.transaction do
+        unless referenced_patient_valid_for_client?(resource, :patient_id) && resource.save(context: :api) && fhir_map.all? { |_k, v| v[:errors].blank? }
+          req_json = JSON.parse(request_body)
+          status_unprocessable_entity(resource, fhir_map, req_json) && (raise SaveError)
+        end
+
+        Rails.logger.info "Created Lab Result (ID: #{resource.id}) for Patient with ID: #{resource.patient_id}"
+        History.lab_result(patient: resource.patient_id,
+                           created_by: @current_actor_label,
+                           comment: "New lab result added via API (ID: #{resource.id}).")
+      end
     else
       status_not_found && return
     end
     status_created(resource.as_fhir) && return
   rescue JSON::ParserError
     status_bad_request(['Invalid JSON in request body'])
+  rescue SaveError
+    nil
+  end
+
+  # Create a set of resources as an atomic action
+  #
+  # Supports (writing): Patient, Observation
+  #
+  # POST /fhir/r4
+  def transaction
+    status_unsupported_media_type && return unless content_type_header?('application/fhir+json')
+
+    # Must have Patient write scopes
+    return if doorkeeper_authorize!(*PATIENT_WRITE_SCOPES)
+
+    # Parse in the FHIR
+    request_body = request.body.read
+    contents = FHIR.from_contents(request_body) unless request_body.blank?
+
+    # Only allow a maximum batch size of MAX_TRANSACTION_ENTRIES
+    if !contents&.entry.nil? && contents.entry.length > MAX_TRANSACTION_ENTRIES
```

Comment:
I like this; probably want to make sure it's mentioned somewhere in the API docs

Reply:
https://github.com/SaraAlert/SaraAlert/pull/870/commits/d27d21a2163d3c613b633349fd3eb75f5fdaccf8